dnsflow_read.py

```diff
@@ -1,125 +1,179 @@
+            raise Exception("Specify only interface or pcap_file")
 
         self.interface = interface
         self.pcap_file = pcap_file
         self.pcap_filter = pcap_filter
         self.stats_only = stats_only
-
-        self._pcap = pcap.pcapObject()
-
-        if self.pcap_file is not None:
-            # XXX dpkt pcap doesn't support filters and there's no way to pass
-            # a gzip fd to pylibpcap. Bummer.
-            self._pcap.open_offline(pcap_file)
+        self.live_capture = bool(not self.pcap_file)
+        self.pcap = None
+
+        if self.pcap_file:
+            self.pcap = get_pcap(self.pcap_file)
+            if pcap_filter != DEFAULT_PCAP_FILTER:
+                msg = "ERROR: Only support default filter {} when reading from PCAP.".format(
+                    DEFAULT_PCAP_FILTER
+                )
+                raise ValueError(msg)
         else:
             # Interface
-            # device, snaplen, promisc, to_ms
-            self._pcap.open_live(interface, 65535, 1, 100)
-        # filter, optimize, netmask
-        self._pcap.setfilter(self.pcap_filter, 1, 0)
-
-    # Iterate over individual dnsflow records (multiple per packet).
-    # Skips stats pkts.
-    def flow_iter(self):
-        for pkt in self.pkt_iter():
-            ts = pkt['header']['timestamp']
-            if 'data' not in pkt:
-                # stats pkt
-                continue
-            for record in pkt['data']:
-                yield ts, record
-
-    # Iterate over dnsflow pkts.
-    def pkt_iter(self):
-        while 1:
-            rv = self._pcap.next()
-            if rv == None:
-                if self.pcap_file is not None:
-                    # eof
-                    break
-                else:
-                    # interface, hit to_ms
-                    continue
-            pktlen, buf, ts = rv
-            pkt, err = process_pkt(self._pcap.datalink(), ts, buf,
-                    stats_only=self.stats_only)
-            if err is not None:
-                print err
-                continue
-            yield pkt
-
-#
+            self.pcap = pcap.pcap(name=interface, snaplen=SNAPLEN, promisc=True, timeout_ms=TIMEOUT)
+            self.pcap.setfilter(self.pcap_filter, optimize=1)
+
+    def iter_interface(self):
+        """ Live-capture packets and process them. """
+        print('Listening on %s: %s' % (self.pcap.name, self.pcap.filter))
+        while True:
+            next_item = next(self.pcap)
+            if next_item is not None:
+                ts, frame = next_item
+                # When live-capturing, have already set the filter, so no need to do additional filtering
+                res = self.handle_frame(ts, frame)
+                if res:
+                    yield res
+
+    def iter_pcap(self):
+        """ Iterate through pcap file and process packets. """
+        for ts, frame in self.pcap:
+            res = self.handle_frame(ts, frame, filter=True)
```

Comment:
When live-capturing on an interface, we can do `self.pcap.setfilter(self.pcap_filter, optimize=1)` (https://github.com/deepfield/dnsflow/pull/6/files#diff-e4ccc09a857a261cbb381bcec8c3f53bR74-R75), but  `dpkt.pcap.Reader(f)`'s `setfilter` method is not implemented https://dpkt.readthedocs.io/en/latest/_modules/dpkt/pcap.html#Reader.setfilter
so I am passing `filter=True` to the `handle_frame` method itself. This will cause default filtering of packets, to only accept UDP packets with dport=5300. This is the same as the default filter used when live-capturing, but there is also capability in live-capturing case to add extra filters. This 'extra filtering' capability used to be supported in both modes (read from pcap and live-capture) and now will just be supported for live-capture.